### PR TITLE
JsonSerializer.Serialize with null type parameter

### LINKED ISSUE #20823: Do not pass null value for parameter `Type inputType` in JsonSerializer.Serialize
##  Do not pass a null for parameter `Type inputType` in JsonSerializer.Serialize(...) overloads

Passing in null for the `Type` parameter of the `JsonSerialaizer.Serialize` is unacceptable and should throw ArgumentNullException in that case. It was throwing the exception in most cases except one edge case: for certain `JsonSerializer.Serialize`, `JsonSerializer.SerializeAsync` and `JsonSerializer.SerializeToUtf8Bytes()` overloads which accepting `object value, Type inputType` as parameters, and when those `value` and `inputType` parameters are both `null` it was producing "null" JSON string. Now it throws `ArgumentNullException` correctly whenever `Type inputType` parameter is passed as null. Related to https://github.com/dotnet/runtime/pull/528#issuecomment-589298591

### Version introduced

5.0

### Old behavior
```csharp
public static void TestSerializeNullType()
{
    // Before (3.1): Returned a string with value "null"
    // Same with other Serialize{Async} overloads
    JsonSerializer.Serialize(null, null);

    // Before (3.1): Returned a byte[] with value "null"
    JsonSerializer.SerializeToUtf8Bytes(null, null);
}
```
### New behavior
Passing in null for the Type parameter while serializing now correctly throws ArgumentNullException rather than returning "null" JSON:
```csharp
public static void TestSerializeNullType()
{
    // After (5.0): ArgumentNullException: Value cannot be null. (Parameter 'inputType')
    // Same with other Serialize{Async} overloads
    JsonSerializer.Serialize(null, null);

    // After (5.0): ArgumentNullException: Value cannot be null. (Parameter 'inputType')
    JsonSerializer.SerializeToUtf8Bytes(null, null);
}
```
### Reason for change

Passing in null for the `Type inputType` parameter of the JsonSerialaizer is unacceptable and should always throw ArgumentNullException. 
 
### Recommended action

Avoid passing null value for `Type inputType` parameter

### Category

- [ ] ASP.NET Core
- [ ] C#
- [ ] Code analysis
- [x] Core .NET libraries
- [ ] Cryptography
- [ ] Data
- [ ] Debugger
- [ ] Deployment
- [ ] Globalization
- [ ] Interop
- [ ] JIT
- [ ] LINQ
- [ ] Managed Extensibility Framework (MEF)
- [ ] MSBuild
- [ ] Networking
- [ ] Printing
- [ ] Security
- [x] Serialization
- [ ] Visual Basic
- [ ] Windows Forms
- [ ] Windows Presentation Foundation (WPF)
- [ ] XML, XSLT

### Affected APIs
[System.Text.Json.JsonSerializer.Serialize(Object, Type, JsonSerializerOptions)](https://docs.microsoft.com/dotnet/api/system.text.json.jsonserializer.serialize#System_Text_Json_JsonSerializer_Serialize_System_Object_System_Type_System_Text_Json_JsonSerializerOptions_)
[System.Text.Json.JsonSerializer.Serialize(Utf8JsonWriter, Object, Type, JsonSerializerOptions)](https://docs.microsoft.com/dotnet/api/system.text.json.jsonserializer.serialize#System_Text_Json_JsonSerializer_Serialize_System_Text_Json_Utf8JsonWriter_System_Object_System_Type_System_Text_Json_JsonSerializerOptions_)
[System.Text.Json.JsonSerializer.SerializeAsync(Stream, Object, Type, JsonSerializerOptions, CancellationToken)](https://docs.microsoft.com/dotnet/api/system.text.json.jsonserializer.serialize#System_Text_Json_JsonSerializer_SerializeAsync_System_IO_Stream_System_Object_System_Type_System_Text_Json_JsonSerializerOptions_System_Threading_CancellationToken_)
[System.Text.Json.JsonSerializer.SerializeToUtf8Bytes(Object, Type, JsonSerializerOptions)](https://docs.microsoft.com/dotnet/api/system.text.json.jsonserializer.serializetoutf8bytes#System_Text_Json_JsonSerializer_SerializeToUtf8Bytes_System_Object_System_Type_System_Text_Json_JsonSerializerOptions_)

---
#### Issue metadata

* Issue type: breaking-change

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #20823.

[Preview link](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/serialization?branch=pr-en-us-20973).

### CODE REVIEW

**Review by tdykstra — APPROVED** [review]
LGTM

File: includes/core-changes/serialization/5.0/jsonserializer-serialize-throws-argumentnullexception-for-null-type.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#### Change description
+
+In .NET Core 3.1, the <xref:System.Text.Json.JsonSerializer.Serialize%2A?displayProperty=nameWithType>, <xref:System.Text.Json.JsonSerializer.SerializeAsync(System.IO.Stream,System.Object,System.Type,System.Text.Json.JsonSerializerOptions,System.Threading.CancellationToken)?displayProperty=nameWithType>, and <xref:System.Text.Json.JsonSerializer.SerializeToUtf8Bytes(System.Object,System.Type,System.Text.Json.JsonSerializerOptions)?displayProperty=nameWithType> overloads that have a <xref:System.Type> parameter throw an <xref:System.ArgumentNullException> when `null` is passed for the `Type inputType` parameter, but not if the `Object value` parameter is also `null`. Starting in .NET 5.0, these methods *always* throw an <xref:System.ArgumentNullException> when `null` is passed for the <xref:System.Type> parameter.
```
Comment (buyaa-n) [inline includes/core-changes/serialization/5.0/jsonserializer-serialize-throws-argumentnullexception-for-null-type.md:7]:
```suggestion
In .NET Core 3.1, the <xref:System.Text.Json.JsonSerializer.Serialize(System.Object,System.Type,System.Text.Json.JsonSerializerOptions)?displayProperty=nameWithType> and <xref:System.Text.Json.JsonSerializer.SerializeToUtf8Bytes(System.Object,System.Type,System.Text.Json.JsonSerializerOptions)?displayProperty=nameWithType> overloads that have a <xref:System.Type> parameter throw an <xref:System.ArgumentNullException> when `null` is passed for the `Type inputType` parameter, but not if the `Object value` parameter is also `null`. Starting in .NET 5.0, these methods *always* throw an <xref:System.ArgumentNullException> when `null` is passed for the <xref:System.Type> parameter.
```

**Review by buyaa-n — APPROVED** [review]
One small suggestion, overall looks good, thanks!

File: includes/core-changes/serialization/5.0/jsonserializer-serialize-throws-argumentnullexception-for-null-type.md (hunk 1 above)
Comment (gewarren) [reply includes/core-changes/serialization/5.0/jsonserializer-serialize-throws-argumentnullexception-for-null-type.md:7]: Why not SerializeAsync?

Comment (buyaa-n) [reply includes/core-changes/serialization/5.0/jsonserializer-serialize-throws-argumentnullexception-for-null-type.md:7]: You could keep the async version if you want, for me, it looks better having only 2 main methods, as SerializeAsync is listed in the affected APIs list anyways.

Comment (ahsonkhan) [reply includes/core-changes/serialization/5.0/jsonserializer-serialize-throws-argumentnullexception-for-null-type.md:7]: Providing more explicit detail on all the APIs were affected can be helpful for someone doing Ctrl+F(ind) for the particular name.

**Pull request merged**